### PR TITLE
JENKINS-37878: Display url provider

### PR DESCRIPTION
**Pull request by imeredith**
# Description
WIP, probably isnt the best place for this stuff.

See [JENKINS-37878](https://issues.jenkins-ci.org/browse/JENKINS-37878).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.
- [ ] Ran Acceptance Test Harness against PR changes.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

@jenkinsci/code-reviewers @reviewbybees

### CODE REVIEW

**Comment by imeredith** [comment]
@vivek where would the best place for this to go be?

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/BlueOceanDisplayURLImpl.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+@Extension
+public class BlueOceanDisplayURLImpl extends DisplayURLProvider {
+    @Override
+    public String getRoot() {
+        Jenkins jenkins = Jenkins.getInstance();
+        if (jenkins == null) {
+            throw new IllegalStateException("Jenkins has not started");
+        }
+        String root = jenkins.getRootUrl();
+        if (root == null) {
```
Comment (i386) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/BlueOceanDisplayURLImpl.java]: Perhaps we should make upstream DisplayURLProvider do this for you if you call `super.getRoot()`?

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/BlueOceanDisplayURLImpl.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        } else {
+            jobPath = encode(project.getFullName());
+        }
+
+        return getRoot() + "organizations/jenkins/pipelines/" + jobPath + "/";
+    }
+
+    private String encode(String path) {
+        try {
+            return URLEncoder.encode(path, "UTF-8");
```
Comment (i386) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/BlueOceanDisplayURLImpl.java]: BTW Jenkins uses `Util.encode` for this. Should we be using this instead?

**Comment by i386** [comment]
@imeredith WDYT of just moving this to its own module?

**Comment by vivek** [comment]
@imeredith rest api implementations are not the best place for display url provider implementation to be. Maybe move it in separate plugin or make it part of blueocean-config or blueocean-dashboard?

**Comment by i386** [comment]
How about a new module 'blueocean-services' or something like that? Then again config is probably just fine

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/BlueOceanDisplayURLImpl.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * Created by ivan on 13/09/16.
+ */
+@Extension
+public class BlueOceanDisplayURLImpl extends DisplayURLProvider {
+    @Override
+    public String getRoot() {
+        Jenkins jenkins = Jenkins.getInstance();
+        if (jenkins == null) {
```
Comment (tfennelly) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/BlueOceanDisplayURLImpl.java]: I don't think this check is required. `Jenkins.getInstance()` is `@Nonnull` return.

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/BlueOceanDisplayURLImpl.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            Job job = ((Job) run.getParent());
+            return getJobURL(job) + "detail/" + encode(job.getDisplayName()) + "/" + run.getNumber() + "/";
+        }
+
+        return null;
+    }
+
+    @Override
+    public String getChangesURL(Run<?, ?> run) {
+        return getRunURL(run) + "changes";
```
Comment (tfennelly) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/BlueOceanDisplayURLImpl.java]: Looks like `getRunURL` can return `null`?

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/BlueOceanDisplayURLImpl.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    private String encode(String path) {
+        try {
+            return URLEncoder.encode(path, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            throw new ServiceException.UnexpectedErrorException("Error encoding url");
+        }
+    }
+    @Override
+    public String getTestUrl(TestResult result) {
+        return getRunURL(result.getRun()) + "/tests";
```
Comment (tfennelly) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/BlueOceanDisplayURLImpl.java]: Ditto earlier wrt `getRunUrl` returning `null`.

File: blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/BlueOceanDisplayURLImplTest.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+            file += "echo '"+msg+"'";
+            return this;
+        }
+
+        public String getFileContents() {
+            return file;
+        }
+    }
+
+}
```
Comment (tfennelly) [inline blueocean-pipeline-api-impl/src/test/java/io/jenkins/blueocean/rest/impl/pipeline/BlueOceanDisplayURLImplTest.java]: Kudos on the tests here man !!

**Comment by michaelneale** [comment]
no please no more modules ;)

**Comment by i386** [comment]
@michaelneale we are using maven 🍘

**Pull request closed without merging**